File: src/main/java/org/jcodec/codecs/mpeg4/MPEG4Decoder.java
package org.jcodec.codecs.mpeg4;

import static org.jcodec.codecs.mpeg4.MPEG4BiRenderer.renderBi;
import static org.jcodec.codecs.mpeg4.MPEG4Bitstream.checkResyncMarker;
import static org.jcodec.codecs.mpeg4.MPEG4Bitstream.readBi;
import static org.jcodec.codecs.mpeg4.MPEG4Bitstream.readCoeffIntra;
import static org.jcodec.codecs.mpeg4.MPEG4Bitstream.readInterCoeffs;
import static org.jcodec.codecs.mpeg4.MPEG4Bitstream.readInterModeCoeffs;
import static org.jcodec.codecs.mpeg4.MPEG4Bitstream.readIntraMode;
import static org.jcodec.codecs.mpeg4.MPEG4Bitstream.readVideoPacketHeader;
import static org.jcodec.codecs.mpeg4.MPEG4Renderer.renderInter;
import static org.jcodec.codecs.mpeg4.MPEG4Renderer.renderIntra;

import java.nio.ByteBuffer;

import org.jcodec.codecs.mpeg4.Macroblock.Vector;
import org.jcodec.common.VideoCodecMeta;
import org.jcodec.common.VideoDecoder;
import org.jcodec.common.io.BitReader;
import org.jcodec.common.model.ColorSpace;
import org.jcodec.common.model.Picture;
import org.jcodec.common.model.Rect;
import org.jcodec.common.model.Size;

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 * 
 * @author The JCodec project
 * 
 */
public class MPEG4Decoder extends VideoDecoder {
    private Picture[] refs = new Picture[2];

    private Macroblock[] prevMBs;
    private Macroblock[] mbs;

    public final static int I_VOP = 0;
    public final static int P_VOP = 1;
    public final static int B_VOP = 2;
    public final static int S_VOP = 3;
    public final static int N_VOP = 4;
    
    private MPEG4DecodingContext ctx;

    @Override
    public Picture decodeFrame(ByteBuffer data, byte[][] buffer) {
        if (ctx == null)
            ctx = new MPEG4DecodingContext();
        if (!ctx.readHeaders(data))
            return null;

        ctx.fcodeForward = ctx.fcodeBackward = ctx.intraDCThreshold = 0;

        BitReader br = BitReader.createBitReader(data);

        if (!ctx.readVOPHeader(br))
            return null;

        mbs = new Macroblock[ctx.mbWidth * ctx.mbHeight];
        for (int i = 0; i < mbs.length; i++) {
            mbs[i] = new Macroblock();
        }

        Picture decoded = null;
        if (ctx.codingType != B_VOP) {
            switch (ctx.codingType) {
            case I_VOP:
                decoded = decodeIFrame(br, ctx, buffer);
                break;
            case P_VOP:
                decoded = decodePFrame(br, ctx, buffer, ctx.fcodeForward);
                break;
            case S_VOP:
                throw new RuntimeException("GMC not supported.");
            case N_VOP:
                return null;
            }

            refs[1] = refs[0];
            refs[0] = decoded;

            prevMBs = mbs;
        } else {
            decoded = decodeBFrame(br, ctx, buffer, ctx.quant, ctx.fcodeForward, ctx.fcodeBackward);
        }
        // We don't want to overread
        br.terminate();

        return decoded;
    }

    private Picture decodeIFrame(BitReader br, MPEG4DecodingContext ctx, byte[][] buffer) {
        Picture p = new Picture(ctx.mbWidth << 4, ctx.mbHeight << 4, buffer, null, ColorSpace.YUV420, 0,
                new Rect(0, 0, ctx.width, ctx.height));
        int bound = 0;

        for (int y = 0; y < ctx.mbHeight; y++) {
            for (int x = 0; x < ctx.mbWidth; x++) {
                Macroblock mb = mbs[y * ctx.mbWidth + x];
                mb.reset(x, y, bound);

                readIntraMode(br, ctx, mb);
                int index = x + y * ctx.mbWidth;
                Macroblock aboveMb = null;
                Macroblock aboveLeftMb = null;
                Macroblock leftMb = null;
                int apos = index - ctx.mbWidth;
                int lpos = index - 1;
                int alpos = index - 1 - ctx.mbWidth;
                if (apos >= bound)
                    aboveMb = mbs[apos];
                if (alpos >= bound)
                    aboveLeftMb = mbs[alpos];
                if (x > 0 && lpos >= bound)
                    leftMb = mbs[lpos];
                readCoeffIntra(br, ctx, mb, aboveMb, leftMb, aboveLeftMb);
                
                x = mb.x;
                y = mb.y;
                bound = mb.bound;

                renderIntra(mb, ctx);
                
                putPix(p, mb, x, y);
            }
        }

        return p;
    }

    Picture decodePFrame(BitReader br, MPEG4DecodingContext ctx, byte[][] buffers, int fcode) {
        int bound = 0;
        int mbWidth = ctx.mbWidth;
        int mbHeight = ctx.mbHeight;

        Picture p = new Picture(ctx.mbWidth << 4, ctx.mbHeight << 4, buffers, null, ColorSpace.YUV420, 0,
                new Rect(0, 0, ctx.width, ctx.height));

        for (int y = 0; y < mbHeight; y++) {
            for (int x = 0; x < mbWidth; x++) {
                // skip stuffing
                while (br.checkNBit(10) == 1)
                    br.skip(10);

                if (checkResyncMarker(br, fcode - 1)) {
                    bound = readVideoPacketHeader(br, ctx, fcode - 1, true, false, true);
                    x = bound % mbWidth;
                    y = bound / mbWidth;
                }

                int index = x + y * ctx.mbWidth;
                Macroblock aboveMb = null;
                Macroblock aboveLeftMb = null;
                Macroblock leftMb = null;
                Macroblock aboveRightMb = null;
                int apos = index - ctx.mbWidth;
                int lpos = index - 1;
                int alpos = index - 1 - ctx.mbWidth;
                int arpos = index + 1 - ctx.mbWidth;
                if (apos >= bound)
                    aboveMb = mbs[apos];
                if (alpos >= bound)
                    aboveLeftMb = mbs[alpos];
                if (x > 0 && lpos >= bound)
                    leftMb = mbs[lpos];
                if (arpos >= bound && x < ctx.mbWidth - 1)
                    aboveRightMb = mbs[arpos];
                
                Macroblock mb = mbs[y * ctx.mbWidth + x];

                mb.reset(x, y, bound);

                readInterModeCoeffs(br, ctx, fcode, mb, aboveMb, leftMb, aboveLeftMb, aboveRightMb);
                
                renderInter(ctx, refs, mb, fcode, 0, false);
                
                putPix(p, mb, x, y);
            }
        }

        return p;
    }

    private Picture decodeBFrame(BitReader br, MPEG4DecodingContext ctx, byte[][] buffers, int quant, int fcodeForward,
            int fcodeBackward) {

        Picture p = new Picture(ctx.mbWidth << 4, ctx.mbHeight << 4, buffers, null, ColorSpace.YUV420, 0,
                new Rect(0, 0, ctx.width, ctx.height));

        Vector pFMV = new Vector(), pBMV = new Vector();
        for (int y = 0; y < ctx.mbHeight; y++) {
            pBMV.x = pBMV.y = pFMV.x = pFMV.y = 0;

            for (int x = 0; x < ctx.mbWidth; x++) {
                Macroblock mb = mbs[y * ctx.mbWidth + x];
                Macroblock lastMB = prevMBs[y * ctx.mbWidth + x];
                final int fcodeMax = (fcodeForward > fcodeBackward) ? fcodeForward : fcodeBackward;

                if (checkResyncMarker(br, fcodeMax - 1)) {
                    int bound = readVideoPacketHeader(br, ctx, fcodeMax - 1, fcodeForward != 0, fcodeBackward != 0,
                            ctx.intraDCThreshold != 0);
                    x = bound % ctx.mbWidth;
                    y = bound / ctx.mbWidth;
                    pBMV.x = pBMV.y = pFMV.x = pFMV.y = 0;
                }
                mb.x = x;
                mb.y = y;

                mb.quant = quant;

                if (lastMB.mode == MPEG4Consts.MODE_NOT_CODED) {
                    mb.cbp = 0;
                    mb.mode = MPEG4Consts.MODE_FORWARD;
                    readInterCoeffs(br, ctx, mb);
                    renderInter(ctx, refs, lastMB, fcodeForward, 1, true);
                    putPix(p, lastMB, x, y);
                    continue;
                }

                readBi(br, ctx, fcodeForward, fcodeBackward, mb, lastMB, pFMV, pBMV);
                renderBi(ctx, refs, fcodeForward, fcodeBackward, mb);
                
                putPix(p, mb, x, y);
            }
        }

        return p;
    }
    
    public static void putPix(Picture p, Macroblock mb, int x, int y) {
        byte[] plane0 = p.getPlaneData(0);
        int dsto0 = ((y << 4)) * p.getWidth() + (x << 4);
        for (int row = 0, srco = 0; row < 16; row++, dsto0 += p.getWidth()) {
            for (int col = 0; col < 16; col++, srco++) {
                plane0[dsto0 + col] = mb.pred[0][srco];
            }
        }
        for (int pl = 1; pl < 3; pl++) {
            byte[] plane = p.getPlaneData(pl);
            int dsto = ((y << 3)) * p.getPlaneWidth(pl) + (x << 3);
            for (int row = 0, srco = 0; row < 8; row++, dsto += p.getPlaneWidth(pl)) {
                for (int col = 0; col < 8; col++, srco++) {
                    plane[dsto + col] = mb.pred[pl][srco];
                }
            }
        }
    }

    @Override
    public VideoCodecMeta getCodecMeta(ByteBuffer data) {
        return VideoCodecMeta.createSimpleVideoCodecMeta(new Size(320, 240), ColorSpace.YUV420J);
    }

    public static int probe(ByteBuffer data) {
        return 0;
    }
}

File: src/main/java/org/jcodec/codecs/mpeg4/MPEG4Decoder.kt
package org.jcodec.codecs.mpeg4

import org.jcodec.codecs.mpeg4.MPEG4BiRenderer.renderBi
import org.jcodec.codecs.mpeg4.Macroblock.Companion.vec
import org.jcodec.common.UsedViaReflection
import org.jcodec.common.VideoCodecMeta
import org.jcodec.common.VideoDecoder
import org.jcodec.common.io.BitReader
import org.jcodec.common.io.BitReader.Companion.createBitReader
import org.jcodec.common.model.ColorSpace
import org.jcodec.common.model.Picture
import org.jcodec.common.model.Rect
import org.jcodec.common.model.Size
import java.nio.ByteBuffer

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 *
 * @author The JCodec project
 */
class MPEG4Decoder : VideoDecoder() {
    private val refs: Array<Picture?> = arrayOfNulls(2)
    private var prevMBs: Array<Macroblock?> = emptyArray()
    private var mbs: Array<Macroblock?> = emptyArray()
    private var ctx: MPEG4DecodingContext? = null
    override fun decodeFrame(data: ByteBuffer, buffer: Array<ByteArray>): Picture? {
        if (ctx == null) ctx = MPEG4DecodingContext()
        val _ctx = ctx!!
        if (!_ctx.readHeaders(data)) return null
        _ctx.intraDCThreshold = 0
        _ctx.fcodeBackward = _ctx.intraDCThreshold
        _ctx.fcodeForward = _ctx.fcodeBackward
        val br = createBitReader(data)
        if (!_ctx.readVOPHeader(br)) return null
        mbs = arrayOfNulls(_ctx.mbWidth * _ctx.mbHeight)
        for (i in mbs.indices) {
            mbs[i] = Macroblock()
        }
        var decoded: Picture? = null
        if (_ctx.codingType != MPEG4Bitstream.B_VOP) {
            when (_ctx.codingType) {
                MPEG4Bitstream.I_VOP -> decoded = decodeIFrame(br, _ctx, buffer)
                MPEG4Bitstream.P_VOP -> decoded = decodePFrame(br, _ctx, buffer, _ctx.fcodeForward)
                MPEG4Bitstream.S_VOP -> throw RuntimeException("GMC not supported.")
                MPEG4Bitstream.N_VOP -> return null
            }
            refs[1] = refs[0]
            refs[0] = decoded!!
            prevMBs = mbs
        } else {
            decoded = decodeBFrame(br, _ctx, buffer, _ctx.quant, _ctx.fcodeForward, _ctx.fcodeBackward)
        }
        // We don't want to overread
        br.terminate()
        return decoded
    }

    private fun decodeIFrame(br: BitReader, ctx: MPEG4DecodingContext, buffer: Array<ByteArray>): Picture {
        val p = Picture(ctx.mbWidth shl 4, ctx.mbHeight shl 4, buffer, null, ColorSpace.YUV420, 0,
                Rect(0, 0, ctx.width, ctx.height))
        var bound = 0
        var y = 0
        while (y < ctx.mbHeight) {
            var x = 0
            while (x < ctx.mbWidth) {
                val mb = mbs[y * ctx.mbWidth + x]
                mb!!.reset(x, y, bound)
                MPEG4Bitstream.readIntraMode(br, ctx, mb)
                val index = x + y * ctx.mbWidth
                var aboveMb: Macroblock? = null
                var aboveLeftMb: Macroblock? = null
                var leftMb: Macroblock? = null
                val apos = index - ctx.mbWidth
                val lpos = index - 1
                val alpos = index - 1 - ctx.mbWidth
                if (apos >= bound) aboveMb = mbs[apos]
                if (alpos >= bound) aboveLeftMb = mbs[alpos]
                if (x > 0 && lpos >= bound) leftMb = mbs[lpos]
                MPEG4Bitstream.readCoeffIntra(br, ctx, mb, aboveMb, leftMb, aboveLeftMb)
                x = mb.x
                y = mb.y
                bound = mb.bound
                MPEG4Renderer.renderIntra(mb, ctx)
                putPix(p, mb, x, y)
                x++
            }
            y++
        }
        return p
    }

    fun decodePFrame(br: BitReader, ctx: MPEG4DecodingContext, buffers: Array<ByteArray>?, fcode: Int): Picture {
        var bound = 0
        val mbWidth = ctx.mbWidth
        val mbHeight = ctx.mbHeight
        val p = Picture(ctx.mbWidth shl 4, ctx.mbHeight shl 4, buffers, null, ColorSpace.YUV420, 0,
                Rect(0, 0, ctx.width, ctx.height))
        var y = 0
        while (y < mbHeight) {
            var x = 0
            while (x < mbWidth) {

                // skip stuffing
                while (br.checkNBit(10) == 1) br.skip(10)
                if (MPEG4Bitstream.checkResyncMarker(br, fcode - 1)) {
                    bound = MPEG4Bitstream.readVideoPacketHeader(br, ctx, fcode - 1, true, false, true)
                    x = bound % mbWidth
                    y = bound / mbWidth
                }
                val index = x + y * ctx.mbWidth
                var aboveMb: Macroblock? = null
                var aboveLeftMb: Macroblock? = null
                var leftMb: Macroblock? = null
                var aboveRightMb: Macroblock? = null
                val apos = index - ctx.mbWidth
                val lpos = index - 1
                val alpos = index - 1 - ctx.mbWidth
                val arpos = index + 1 - ctx.mbWidth
                if (apos >= bound) aboveMb = mbs[apos]
                if (alpos >= bound) aboveLeftMb = mbs[alpos]
                if (x > 0 && lpos >= bound) leftMb = mbs[lpos]
                if (arpos >= bound && x < ctx.mbWidth - 1) aboveRightMb = mbs[arpos]
                val mb = mbs[y * ctx.mbWidth + x]
                mb!!.reset(x, y, bound)
                MPEG4Bitstream.readInterModeCoeffs(br, ctx, fcode, mb, aboveMb, leftMb, aboveLeftMb, aboveRightMb)
                MPEG4Renderer.renderInter(ctx, refs, mb, fcode, 0, false)
                putPix(p, mb, x, y)
                x++
            }
            y++
        }
        return p
    }

    private fun decodeBFrame(br: BitReader, ctx: MPEG4DecodingContext, buffers: Array<ByteArray>, quant: Int, fcodeForward: Int,
                             fcodeBackward: Int): Picture {
        val p = Picture(ctx.mbWidth shl 4, ctx.mbHeight shl 4, buffers, null, ColorSpace.YUV420, 0,
                Rect(0, 0, ctx.width, ctx.height))
        val pFMV = vec()
        val pBMV = vec()
        //To prevent unexpected behaviour in Javascript, final variables must be declared at method level and not inside loops
        val fcodeMax = if (fcodeForward > fcodeBackward) fcodeForward else fcodeBackward
        var y = 0
        while (y < ctx.mbHeight) {
            pFMV.y = 0
            pFMV.x = pFMV.y
            pBMV.y = pFMV.x
            pBMV.x = pBMV.y
            var x = 0
            while (x < ctx.mbWidth) {
                val mb = mbs[y * ctx.mbWidth + x]
                val lastMB = prevMBs[y * ctx.mbWidth + x]
                if (MPEG4Bitstream.checkResyncMarker(br, fcodeMax - 1)) {
                    val bound = MPEG4Bitstream.readVideoPacketHeader(br, ctx, fcodeMax - 1, fcodeForward != 0, fcodeBackward != 0,
                            ctx.intraDCThreshold != 0)
                    x = bound % ctx.mbWidth
                    y = bound / ctx.mbWidth
                    pFMV.y = 0
                    pFMV.x = pFMV.y
                    pBMV.y = pFMV.x
                    pBMV.x = pBMV.y
                }
                mb!!.x = x
                mb.y = y
                mb.quant = quant
                if (lastMB!!.mode == MPEG4Consts.MODE_NOT_CODED) {
                    mb.cbp = 0
                    mb.mode = MPEG4Consts.MODE_FORWARD
                    MPEG4Bitstream.readInterCoeffs(br, ctx, mb)
                    MPEG4Renderer.renderInter(ctx, refs, lastMB, fcodeForward, 1, true)
                    putPix(p, lastMB, x, y)
                    x++
                    continue
                }
                MPEG4Bitstream.readBi(br, ctx, fcodeForward, fcodeBackward, mb, lastMB, pFMV, pBMV)
                renderBi(ctx, refs as Array<Picture>, fcodeForward, fcodeBackward, mb)
                putPix(p, mb, x, y)
                x++
            }
            y++
        }
        return p
    }

    override fun getCodecMeta(data: ByteBuffer): VideoCodecMeta? {
        val ctx = MPEG4DecodingContext.readFromHeaders(data.duplicate()) ?: return null
        return VideoCodecMeta.createSimpleVideoCodecMeta(Size(ctx.width, ctx.height), ColorSpace.YUV420J)
    }

    companion object {
        fun putPix(p: Picture, mb: Macroblock?, x: Int, y: Int) {
            val plane0 = p.getPlaneData(0)
            var dsto0 = (y shl 4) * p.width + (x shl 4)
            var row = 0
            var srco = 0
            while (row < 16) {
                var col = 0
                while (col < 16) {
                    plane0[dsto0 + col] = mb!!.pred[0][srco]
                    col++
                    srco++
                }
                row++
                dsto0 += p.width
            }
            for (pl in 1..2) {
                val plane = p.getPlaneData(pl)
                var dsto = (y shl 3) * p.getPlaneWidth(pl) + (x shl 3)
                var row = 0
                var srco = 0
                while (row < 8) {
                    var col = 0
                    while (col < 8) {
                        plane[dsto + col] = mb!!.pred[pl][srco]
                        col++
                        srco++
                    }
                    row++
                    dsto += p.getPlaneWidth(pl)
                }
            }
        }

        @UsedViaReflection
        fun probe(data: ByteBuffer): Int {
            val ctx = MPEG4DecodingContext.readFromHeaders(data.duplicate()) ?: return 0
            return Math.min(if (ctx.width > 320) if (ctx.width < 1280) 100 else 50 else 50,
                    if (ctx.height > 240) if (ctx.height < 720) 100 else 50 else 50)
        }
    }

}